designdocs/user-handles.md

```diff
@@ -0,0 +1,308 @@
+    Wallet-->>App: Return signed payload
+    App->>RPC: create_msa_handle(..., handle, suffix)*
+    RPC->>Chain: Submit transaction
+    Chain->>Chain: Check validity and availability
+    Chain-->>App: MsaHandleCreated event with handle
+    App-->>RPC: Publish Profile Announcement/batch with (canonical) DSNP handle
+    App->>User: Proceed with setup
+    User->>App: Request to get MSA ID for handle
+    App->>RPC: get_msa_id_for_handle(handle)
+    RPC->>Chain: Query handle mapping state (State Query MSA<->Handle)
+    Chain-->>RPC: Return handle mapping state (State Query MSA<->Handle)
+    RPC-->>App: Return MSA ID for handle
+    User->>App: Request to get handle for MSA ID
+    App->>RPC: get_handle_for_msa_id(msa_id)
+    RPC->>Chain: Query MSA ID mapping state (Reverse Index Handle<->MSA)
+    Chain-->>RPC: Return MSA ID mapping state (Reverse Index Handle<->MSA)
+    RPC-->>App: Return handle for MSA ID
+```
+
+## Storage
+
+* **HandleSuffixMin**: This storage value (```u32``) will keep track of the minimum suffix value allowed for a given handle.
+* **HandleSuffixMax**: This storage value (```u32``) will keep track of the maximum suffix value allowed for a given handle.
+* **HandleRetirementPeriod**: This storage value (```u32``) will keep track of the number of blocks a handle must be retired before it can be reused.
+* **MSAIdToUserHandle**: This storage will map MSAIds to their corresponding user handles.
+* **UserHandlesToMSAId**: This storage will map user handles to their corresponding MSAIds.
+* **RetiredHandles**: This storage will keep track of all retired handles and the block number at which they were retired. This probably is a spike to run few ideas on how to implement this.
+* **UsedSuffixes**: This storage will keep track of all used suffixes to ensure that no two handles have the same suffix. (May not be necessary)
+* **Seed (current and previous)**: This storage values (```tuple (s0, s1)```) will keep track of the current and previous seed values.
+
+## Required Extrinsics
+
+### Primitives
+
+``` rust
+
+MsaHandlePayload {
+    handle: &[u8],
+    suffix: u32,
+    msa_id: u64,
+}
+```
+
+### Create user handle with chosen suffix
+
+ As a network frequency should allow users to choose their own handle and suffix. This extrinsic will allow users to create a handle with a chosen suffix.
+
+``` rust
+Input
+
+* origin - must be a signed origin
+* owner msa id - the MSA ID of the user. MSA ID must be created before calling this extrinsic.
+* proof - the proof of ownership of the handle, ```MsaHandlePayload``` signed by the user/provider private key
+
+Output
+* Event - `MsaHandleCreated` with the MSA ID and the handle
+
+* Errors - 
+    * `HandleAlreadyExists` if the handle already exists
+    * `InvalidMsaHandle` if the handle is invalid
+    * `Unauthorized` if the signature is invalid
+    * `InvalidSuffix` if the suffix is invalid
+    * `SuffixAlreadyExists` if the suffix already exists
```

Comment:
I don't think `SuffixAlreadyExists` is a needed error. We can just say handle not available to refer to the handle+suffix.

Reply:
Yep not needed now, however since we are doing optional suffix, what if someone uses extrinsic to claim some handle and suffix without checking if that suffix will be available or not

So not sure if its useful to send an error stating exactly same, but yes user can always check range of allowed suffixes.